File: Baekjoon/Graph/[11123]양 한마리．．． 양 두마리．．．/solution.py
def bfs(m, v, x, y, H, W):
    dir = [[1, 0], [0, 1], [-1, 0], [0, -1]]
    q = []

    q.append([x, y])
    v[x][y] = True

    while len(q) != 0:
        cx, cy = q.pop(0)

        for i in range(len(dir)):
            nx = cx + dir[i][0]
            ny = cy + dir[i][1]

            if (nx < 0 or ny < 0 or nx >= H or ny >= W):
                continue

            if((visited[nx][ny] == False) and (matrix[nx][ny] == '#')):
                q.append([nx, ny])
                visited[nx][ny] = True


T = int(input())

for t in range(T):
    H, W = map(int, input().split())
    res = 0

    visited = [[False for i in range(W)] for j in range(H)]
    matrix = ["" for i in range(H)]

    for i in range(H):
        matrix[i] = input()

    for i in range(len(matrix)):
        for j in range(len(matrix[i])):
            if ((visited[i][j] == False) and (matrix[i][j] == '#')):
                bfs(matrix, visited, i, j, H, W)
                res += 1

    print(res)

File: Baekjoon/Graph/[11123]양 한마리．．． 양 두마리．．．/solution.cpp
#include <iostream>
#include <queue>
#include <string>
#include <vector>

using namespace std;

#define endl "\n"

typedef pair<int, int> pii;

vector<vector<int>> matrix;
vector<vector<bool>> visited;
vector<pii> dir;
int T, H, W, res;

void bfs(int x, int y);

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  cin >> T;

  dir = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

  for (; T > 0; T--) {
    cin >> H >> W;

    matrix = vector<vector<int>>(H, vector<int>(W));
    visited = vector<vector<bool>>(H, vector<bool>(W, false));
    res = 0;

    for (int i = 0; i < H; i++) {
      string row;

      cin >> row;
      for (int j = 0; j < W; j++) {
        matrix[i][j] = row[j];
      }
    }

    for (int i = 0; i < H; i++) {
      for (int j = 0; j < W; j++) {
        if (!visited[i][j] && matrix[i][j] == '#') {
          bfs(i, j);
          res++;
        }
      }
    }

    cout << res << endl;
  }

  return 0;
}

void bfs(int x, int y) {
  queue<pii> q;

  q.push({x, y});
  visited[x][y] = true;

  while (!q.empty()) {
    int cx = q.front().first;
    int cy = q.front().second;

    q.pop();

    for (int i = 0; i < dir.size(); i++) {
      int nx = cx + dir[i].first;
      int ny = cy + dir[i].second;

      if (nx < 0 || ny < 0 || nx >= H || ny >= W) continue;

      if (!visited[nx][ny] && matrix[nx][ny] == '#') {
        q.push({nx, ny});
        visited[nx][ny] = true;
      }
    }
  }
}